misk/src/main/kotlin/misk/eventrouter/RealEventRouter.kt

```diff
@@ -4,80 +4,170 @@ import com.google.common.cache.CacheBuilder
+          return clusterConnector.connectSocket(hostname, webSocketListener)
         }
       })
 
-  val webSocketListener = object : WebSocketListener() {
-    override fun onMessage(webSocket: WebSocket, text: String) {
-      val event = eventJsonAdapter.fromJson(text)!!
-      when (event) {
-        is SocketEvent.Subscribe -> remoteSubscribers.put(event.topic, webSocket)
-
-        is SocketEvent.Event -> {
-          localSubscribers.get(event.topic).forEach {
-            it.enqueue(executor, LocalSubscription.Action.Event(event.message))
-          }
+  sealed class Action {
+    data class OnMessage(val webSocket: WebSocket, val text: String) : Action()
+    data class ClusterChanged(val newSnapshot: ClusterSnapshot) : Action()
+    data class Publish(val topic: String, val event: Any) : Action()
+    data class Subscribe(val localSubscription: LocalSubscription<*>) : Action()
+    data class CancelSubscription(val localSubscription: LocalSubscription<*>) : Action()
+    class LeaveCluster : Action()
+  }
 
-          remoteSubscribers.get(event.topic).forEach {
-            it.send(eventJsonAdapter.toJson(event))
-          }
-        }
-      }
+  private val webSocketListener = object : WebSocketListener() {
+    override fun onMessage(webSocket: WebSocket, text: String) {
+      enqueue(Action.OnMessage(webSocket, text))
     }
   }
 
   private val topicPeer = object : TopicPeer {
     override fun acceptWebSocket(webSocket: WebSocket): WebSocketListener = webSocketListener
 
     override fun clusterChanged(clusterSnapshot: ClusterSnapshot) {
-      this@RealEventRouter.clusterSnapshot = clusterSnapshot
-      initCluster()
+      if (hasClusterSnapshot.compareAndSet(false, true)) {
+        this@RealEventRouter.clusterSnapshot = clusterSnapshot
+        logger.debug { "[${clusterSnapshot.self}]: cluster changed: $clusterSnapshot" }
+        executor.submit({ drainQueue() })
+      } else {
+        enqueue(Action.ClusterChanged(clusterSnapshot))
+      }
     }
   }
 
-  /** We queue up subscribes when we don't have a cluster. */
-  private fun initCluster() {
-    while (subscriptionQueue.isNotEmpty()) {
-      val queuedSubscribe = subscriptionQueue.pop()
-      subscribe(queuedSubscribe)
-    }
+  internal fun drainQueue() {
+    if (!hasClusterSnapshot.get()) return
+
+    while (true) {
+      val action = actionQueue.poll() ?: return
 
-    while (publishQueue.isNotEmpty()) {
-      val (topic, event) = publishQueue.pop()
-      publish(topic, event)
+      logger.debug { "[${clusterSnapshot.self}]: processing action: $action" }
+      when (action) {
+        is Action.LeaveCluster -> clusterConnector.leaveCluster(topicPeer)
+
+        is Action.ClusterChanged -> {
+          val removedHosts = clusterSnapshot.hosts.minus(action.newSnapshot.hosts)
+          for (host in removedHosts) {
+            logger.debug { "[${clusterSnapshot.self}]: invalidating $host" }
+            hostToSocket.invalidate(host)
```

Comment:
this isn’t sufficient . . . we need to shutdown the socket right? As-is I think we’re leaking a websocket

Reply:
yeah this whole block needs to be handled better. Should have just put a TODO()